Dockerfile

```diff
@@ -1,40 +1,78 @@
-FROM --platform=linux/amd64 ruby:3.0.6
-RUN rm /bin/sh && ln -s /bin/bash /bin/sh
-WORKDIR /usr/src/app
-EXPOSE 3100
+ARG RUBY_VERSION=3.0.6
+FROM registry.docker.com/library/ruby:$RUBY_VERSION-slim as base
 
-RUN apt-get update -y && apt-get install -y --no-install-recommends \
-  default-jre \
-  default-mysql-client \
-  imagemagick \
-  libprotobuf-dev \
-  protobuf-compiler 
+# Rails app lives here
+WORKDIR /rails
 
-# Workaround for PhantomJS: https://github.com/DMOJ/online-judge/pull/1270
-ENV OPENSSL_CONF /etc/ssl/
+# Install base packages
+RUN apt-get update -qq && \
+    apt-get install --no-install-recommends -y \
+    curl libjemalloc2 libcurl4-openssl-dev default-libmysqlclient-dev
 
-ENV NODE_VERSION 16.18.1
-ENV NVM_DIR=/root/.nvm
+# Set production environment
+ENV RAILS_ENV="production" \
+    BUNDLE_DEPLOYMENT="1" \
+    BUNDLE_PATH="/usr/local/bundle" \
+    BUNDLE_WITHOUT="development"
 
+# Install JavaScript dependencies | Needed in base due coffee-rails gem
+ARG NODE_VERSION=16.20.2
+ARG YARN_VERSION=latest
+ENV PATH=/usr/local/node/bin:$PATH
+RUN curl -sL https://github.com/nodenv/node-build/archive/master.tar.gz | tar xz -C /tmp/ && \
+    /tmp/node-build-master/bin/node-build "${NODE_VERSION}" /usr/local/node
+
+# Throw-away build stage to reduce size of final image
+FROM base as build
+
+# Install packages needed to build gems and node modules
+RUN apt-get install --no-install-recommends -y build-essential git libvips pkg-config
+
+# Install JavaScript dependencies
+RUN npm install -g yarn@$YARN_VERSION && rm -rf /tmp/node-build-master
+
+# Install application gems
+COPY Gemfile Gemfile.lock ./
+RUN bundle install && \
+    rm -rf ~/.bundle/ "${BUNDLE_PATH}"/ruby/*/cache "${BUNDLE_PATH}"/ruby/*/bundler/gems/*/.git && \
+    bundle exec bootsnap precompile --gemfile
+
+# Install node modules
 COPY package.json yarn.lock ./
+RUN yarn install --frozen-lockfile
+
+# Copy application code
+COPY . .
+
+# Precompile bootsnap code for faster boot times
+RUN bundle exec bootsnap precompile app/ lib/
+
+# Precompiling assets for production without requiring secret RAILS_MASTER_KEY
+RUN SECRET_KEY_BASE=1 ./bin/rails assets:precompile
```

Comment:
We may want to explicitly state production environment here. Maybe something like this: 

RUN SECRET_KEY_BASE=1 RAILS_ENV=production bundle exec rails assets:precompile && \
    bundle exec bootsnap precompile --gemfile

Reply:
The `RAILS_ENV` variable is already se in [line 13](https://github.com/GSA/search-gov/pull/1528/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R13). No need to duplicated it in here.